src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2222,14 +2219,15 @@ class ShenandoahUpdateHeapRefsTask : public WorkerTask {
 
 void ShenandoahHeap::update_heap_references(bool concurrent) {
   assert(!is_full_gc_in_progress(), "Only for concurrent and degenerated GC");
-
+  ShenandoahRegionIterator update_refs_iterator(this);
   if (concurrent) {
-    ShenandoahUpdateHeapRefsTask<true> task(&_update_refs_iterator);
+    ShenandoahUpdateHeapRefsTask<true> task(&update_refs_iterator);
     workers()->run_task(&task);
   } else {
-    ShenandoahUpdateHeapRefsTask<false> task(&_update_refs_iterator);
+    ShenandoahUpdateHeapRefsTask<false> task(&update_refs_iterator);
     workers()->run_task(&task);
   }
+  assert(cancelled_gc() || !update_refs_iterator.has_next(), "Should have finished update references");
```

Comment:
Help me out here... Can a degenerate GC never pick up in the midst of an update_refs phase, continuing work left from the point of interruption without having to start from scratch?

Reply:
Yes! It can! Good catch! I'll abandon this PR and fix up `shenandoah:master`.